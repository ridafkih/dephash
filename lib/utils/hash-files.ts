import { createHash } from "crypto";
import { readFileSync } from "fs";
import { join } from "path";

/**
 * Hashes an array of file paths contents as a hash.
 * @param paths An array of paths relative to the current working directory.
 * @param rootDirectory Root directory to append to the beginning of paths provided.
 * @returns A hex-representation of the hash.
 */
export const hashFileContents = (paths: string[], rootDirectory?: string) => {
  const hash = createHash("sha1");

  for (const path of paths) {
    const fullPath = rootDirectory ? join(rootDirectory, path) : path;

    try {
      const contents = readFileSync(fullPath);
      hash.update(path).update(contents);
    } catch (error: unknown) {
      if (!error || typeof error !== "object" || !("code" in error)) continue;
      console.warn(
        `There was a(n) ${error.code} error scanning '${path}', which may result in an unexpected hash.`
      );
    }
  }

  return hash.digest("hex").substring(0, 32);
};
