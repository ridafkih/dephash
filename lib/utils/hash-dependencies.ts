import { Platform } from "enums/platform";
import { sync as globSync } from "fast-glob";
import { hashFileContents } from "utils/hash-files";
import type { HashDependenciesOptions } from "types/hash-dependencies-options";

import {
  PACKAGE_JSON_PATTERN,
  IOS_PACKAGE_PATTERN,
  ANDROID_PACKAGE_PATTERN,
  EXPO_CONFIGS_PATTERN,
  PLUGIN_CONFIGS_PATTERN,
} from "constants/glob-patterns";

const getMatchingGlobPaths = (
  pattern: string | string[],
  rootDirectory?: string
) => {
  return globSync(pattern, { cwd: rootDirectory, onlyFiles: true });
};

/**
 * Hashes the dependencies in the current working directory as a string.
 * @param options Optional options to change the behaviour of the dependency hashing functionality.
 * @returns A hash as a string representing the dependencies
 */
export const hashDependencies = ({
  rootDirectory,
  excludePlatforms,
  excludeExpoConfig,
  factorAllDependencyChanges,
  additionalPatterns,
}: HashDependenciesOptions = {}) => {
  const patterns: string[] = [];

  if (factorAllDependencyChanges) {
    const files = getMatchingGlobPaths(PACKAGE_JSON_PATTERN, rootDirectory);
    const hash = hashFileContents(files, { rootDirectory });
    return { hash, fileCount: files.length };
  }

  if (!excludePlatforms?.includes(Platform.Ios))
    patterns.push(IOS_PACKAGE_PATTERN);

  if (!excludePlatforms?.includes(Platform.Android))
    patterns.push(ANDROID_PACKAGE_PATTERN);

  if (!excludeExpoConfig)
    patterns.push(EXPO_CONFIGS_PATTERN, PLUGIN_CONFIGS_PATTERN);

  if (additionalPatterns) patterns.push(...additionalPatterns);

  const files = getMatchingGlobPaths(patterns, rootDirectory);

  const hash = hashFileContents(files, { rootDirectory });

  return { hash, fileCount: files.length };
};
