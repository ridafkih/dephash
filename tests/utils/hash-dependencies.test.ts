import { hashDependencies } from "utils/hash-dependencies";
import { join } from "path";
import { describe, it, expect } from "vitest";

describe("hashFileContents", () => {
  it("should generate a hash given a root folder", () => {
    const { hash, fileCount } = hashDependencies({
      rootDirectory: join(process.cwd(), "tests", "__project__"),
      additionalPatterns: ["**/*.txt"],
    });

    expect(fileCount).toBe(3);
    expect(hash).toBeTypeOf("string");
  });
  it("should always generate the same hash", () => {
    const { hash } = hashDependencies({
      rootDirectory: join(process.cwd(), "tests", "__project__"),
      additionalPatterns: ["**/*.txt"],
    });

    expect(hash).toMatchSnapshot();
  });
});
