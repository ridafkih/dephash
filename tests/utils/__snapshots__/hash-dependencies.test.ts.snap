// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hashFileContents > should always generate the same hash 1`] = `"f10dbe265c75db7f52ca314d7957414b"`;
